velox/functions/sparksql/specialforms/AtLeastNNonNulls.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/AtLeastNNonNulls.h"
+#include "velox/expression/SpecialForm.h"
+
+using namespace facebook::velox::exec;
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+class AtLeastNNonNullsExpr : public SpecialForm {
+ public:
+  AtLeastNNonNullsExpr(
+      TypePtr type,
+      std::vector<ExprPtr>&& inputs,
+      bool inputsSupportFlatNoNullsFastPath)
+      : SpecialForm(
+            std::move(type),
+            std::move(inputs),
+            AtLeastNNonNullsCallToSpecialForm::kAtLeastNNonNulls,
+            inputsSupportFlatNoNullsFastPath,
+            false /* trackCpuUsage */) {}
+
+  void evalSpecialForm(
+      const SelectivityVector& rows,
+      EvalCtx& context,
+      VectorPtr& result) override {
+    context.ensureWritable(rows, type(), result);
+    (*result).clearNulls(rows);
+    auto flatResult = result->asFlatVector<bool>();
+    auto values = flatResult->mutableValues(rows.end())->asMutable<uint64_t>();
+    // Initialize result to all false.
+    bits::andWithNegatedBits(
+        values, rows.asRange().bits(), rows.begin(), rows.end());
+
+    LocalSelectivityVector activeRowsHolder(context, rows);
+    auto activeRows = activeRowsHolder.get();
+    VELOX_DCHECK(activeRows != nullptr);
+    // Load 'n' from the first input.
+    VectorPtr vector;
+    inputs_[0]->eval(*activeRows, context, vector);
+    VELOX_USER_CHECK(
+        vector->isConstantEncoding(),
+        "The first input should be constant encoding.");
+    auto constVector = vector->asUnchecked<ConstantVector<int32_t>>();
+    VELOX_USER_CHECK(
+        !constVector->isNullAt(0), "The first parameter should not be null.");
+    int32_t n = vector->asUnchecked<ConstantVector<int32_t>>()->valueAt(0);
+
+    // If 'n' > inputs_.size() - 1, result should be all false.
+    if (n > inputs_.size() - 1) {
+      return;
+    }
+    // If 'n' <= 0, set result to all true.
+    if (n <= 0) {
+      bits::orBits(values, rows.asRange().bits(), rows.begin(), rows.end());
+      return;
+    }
```

Comment:
How about handling 'n <=0' before the result initialization to avoid setting result to all-false then all-true?

Reply:
Sure, updated.